### PR TITLE
feat: SDK Automatic Push Behavior Integration (rel/1.0.0)

### PR DESCRIPTION
**Pull request by glenn-klaviyo**
## Description
https://github.com/klaviyo/klaviyo-expo-plugin/pull/119
https://github.com/klaviyo/klaviyo-expo-plugin/pull/120
https://github.com/klaviyo/klaviyo-expo-plugin/pull/121

## Summary by CodeRabbit

* **New Features**
  * Added automatic push-token forwarding options for Android and iOS.
  * Android push-open tracking is now automatic by default through `automaticPushOpenTracking`.
  * Improved notification handling to prevent duplicate completion callbacks.

* **Bug Fixes**
  * Removed legacy Android activity modifications and improved notification delegation.

* **Documentation**
  * Added v1.0.0 migration guidance, configuration examples, and upgrade notes.

* **Breaking Changes**
  * Replaced Android `openTracking` with `automaticPushOpenTracking`.

### CODE REVIEW

File: MIGRATION_GUIDE.md [hunk 1]
```diff
@@ -2,6 +2,56 @@
...
+new property enables tracking rather than disabling it:
+
+```js
+["klaviyo-expo-plugin", {
+  "android": { "automaticPushOpenTracking": false }
+}]
+```
+
+> **Note:** If your `MainActivity` still contains `// @generated begin klaviyo-` blocks from a
+> previous prebuild, run `expo prebuild --clean` once to regenerate clean native files.
```
[the remark this answers is not in the document]
Comment (glenn-klaviyo) [reply MIGRATION_GUIDE.md:51]: Not taking this one — the note is deliberately scoped narrower than the concern.

It only addresses apps that still have `// @generated begin klaviyo-` blocks in `MainActivity`. Those blocks exist *because* a previous `expo prebuild` wrote them, which means the reader is in a CNG workflow where `android/` and `ios/` are generated output (gitignored by default in a managed project). For that audience, `--clean` regenerating them is the intended outcome rather than a destructive one.

An app with hand-maintained native directories wouldn't have plugin-generated blocks to clean up, so it isn't the audience for this line. Adding a back-up-first / `patch-project` caveat would expand a one-line pointer into guidance for a workflow this plugin doesn't target.

File: plugin/types/index.ts [hunk 2]
```diff
@@ -23,11 +25,11 @@ export interface KlaviyoPluginProps {
 
 export interface KlaviyoPluginAndroidProps extends KlaviyoPluginAndroidBaseProps {
   logLevel: number;
-  openTracking: boolean;
   notificationIconFilePath: string | undefined;
   notificationColor: string | undefined;
   geofencingEnabled: boolean;
   formsEnabled: boolean;
+  automaticPushOpenTracking: boolean;
```
[the remark this answers is not in the document]
Comment (glenn-klaviyo) [reply plugin/types/index.ts:32]: Analysis is correct, including the conflict with the `accepts undefined (optional)` case in `tests/validateConfig.test.ts` — confirmed that an explicit `undefined` resolves to `undefined` and the writer's `=== true` check then omits the metadata.

Deliberately not changing it. `automaticPushOpenTracking` is a boolean prop; the supported inputs are `true`, `false`, or omission. Passing an explicit `undefined` isn't a configuration we're supporting, so normalizing it in the merge would add a code path for an input that shouldn't occur.

Worth recording the alternative, though: if we do tighten this, the coherent fix is having `validateAndroidConfig` **reject** `undefined` so it fails loudly, rather than silently resolving it to the default. Today it neither defaults nor rejects, and that's the actual inconsistency you've identified. Leaving it as a possible follow-up rather than altering merge semantics on a release-bound PR.

File: tests/validateConfig.test.ts [hunk 3]
```diff
@@ -118,5 +162,23 @@ describe('validateConfig', () => {
...
+      });
+
+      it('accepts undefined (optional)', () => {
+        expect(() => validateIosConfig({ automaticPushTokenForwarding: undefined })).not.toThrow();
+      });
+
+      it('rejects non-boolean values', () => {
+        expect(() => validateIosConfig({ automaticPushTokenForwarding: 1 as unknown as boolean })).toThrow('iOS automaticPushTokenForwarding must be a boolean');
+      });
+    });
```
[the remark this answers is not in the document]
Comment (glenn-klaviyo) [reply tests/validateConfig.test.ts:182]: Accurate about the type contract, but not fixing it here.

`tsconfig.json` sets `include: ["plugin/**/*"]`, so `tests/**` is outside tsc's boundary entirely — CI's `lint-and-type-check` job runs `npx tsc --noEmit` plus a `grep` for `console.log`/`TODO` scoped to `plugin/`, and never runs ESLint. So these fixtures produce no error in any gate.

It's also the file's established pattern rather than something new here. Batching it with the broader test-suite hardening instead of a partial fix that leaves the other 40+ instances inconsistent.

File: tests/withKlaviyoAndroid.gradleProperties.test.ts [hunk 4]
```diff
@@ -56,7 +56,7 @@ describe('Gradle Properties Plugins', () => {
 
     it('defaults to false when geofencingEnabled is undefined', () => {
       const config: any = { modResults: [] };
-      const props = { logLevel: 1, openTracking: true, notificationIconFilePath: undefined, notificationColor: undefined, formsEnabled: true } as KlaviyoPluginAndroidProps;
+      const props = { logLevel: 1, notificationIconFilePath: undefined, notificationColor: undefined, formsEnabled: true } as KlaviyoPluginAndroidProps;
```
[the remark this answers is not in the document]
Comment (glenn-klaviyo) [reply tests/withKlaviyoAndroid.gradleProperties.test.ts:59]: Same disposition as the `validateConfig.test.ts` thread — valid observation, deferring.

The incomplete `as KlaviyoPluginAndroidProps` casts are pre-existing on `master` in both this file and `withKlaviyoAndroid.integration.test.ts`, and `tests/**` is outside `tsconfig.json`'s `include: ["plugin/**/*"]`, so they aren't type-checked in CI.

Routing `mergeAndroidProps` through these fixtures is a reasonable improvement, but it changes what the tests exercise (defaults-applied objects rather than the literal partials), so it belongs in a deliberate test change rather than alongside a docs/props PR.

File: tests/withKlaviyoAndroid.internal.test.ts [hunk 5]
```diff
@@ -1132,6 +688,367 @@ public class MainActivity extends ReactActivity {
...
+    });
+
+    describe('automaticPushOpenTracking flag', () => {
+      const OPEN_TRACKING_KEY = 'com.klaviyo.push.automatic_push_open_tracking';
+
+      // The tests below drive mutateAndroidManifest directly, so they see whatever value is
+      // passed. These two cover what a real app config resolves to via mergeAndroidProps, which
+      // is what determines the flag's actual default.
+      describe('resolved plugin defaults', () => {
+        const { mergeAndroidProps } = require('../plugin/types');
```
[the remark this answers is not in the document]
Comment (glenn-klaviyo) [reply tests/withKlaviyoAndroid.internal.test.ts:770]: Not converting these, and the load order is the reason rather than convention.

This file's `jest.mock('fs')` / `jest.mock('path')` calls are hoisted, and the deferred `require()` calls resolve the modules *after* those mocks are established. Promoting them to static top-level imports risks changing which implementation the tests bind to, i.e. quietly weakening the tests it's meant to tidy.

That's also why it's the dominant pattern here rather than an oversight: `master` has 29 `require()` calls across the test suite, 17 of them in this file. ESLint isn't run in CI (no `lint` script, no ESLint step in `ci.yml`), so the reported violations aren't gating.

Happy to revisit as part of a single test-suite modernization that can verify the mocks still bind correctly.

File: tests/withKlaviyoIos.test.ts [hunk 6]
```diff
@@ -75,6 +75,54 @@ describe('withKlaviyoIos', () => {
       expect(modifiedConfig.modResults.klaviyo_app_group).toBe('group.com.test.app.KlaviyoNotificationServiceExtension.shared');
       expect(modifiedConfig.modResults.klaviyo_badge_autoclearing).toBe(true);
     });
+
+    describe('automaticPushTokenForwarding', () => {
+      it('should not write klaviyo_automatic_push_token_forwarding when prop is omitted', () => {
+        const modifiedConfig = withKlaviyoIos(mockConfig, mockProps) as any;
+        expect(modifiedConfig.modResults.klaviyo_automatic_push_token_forwarding).toBeUndefined();
```
[the remark this answers is not in the document]
Comment (glenn-klaviyo) [reply tests/withKlaviyoIos.test.ts:82]: Agreed, and thanks for carrying the `tsconfig.json` constraint forward — that's the right read. `tests/**` sits outside `include: ["plugin/**/*"]`, and `as any` appears 46 times across the test suite on `master`, so a targeted fix here would be inconsistent with the rest.

Resolving as an optional follow-up, as you suggested.

**Review by belleklaviyo — APPROVED** [review]
